internal/controllers/gateway/route_utils.go

```diff
@@ -129,123 +135,121 @@ func getSupportedGatewayForRoute[T types.RouteT](ctx context.Context, mgrc clien
 			return nil, fmt.Errorf("failed to retrieve gatewayclass for gateway: %w", err)
 		}
 
-		// if the GatewayClass matches this controller we're all set and this controller
-		// should reconcile this object.
-		if gatewayClass.Spec.ControllerName == ControllerName {
-			allowedNamespaces := make(map[string]interface{})
-			var (
-				// set true if we find any AllowedRoutes. there may be none, in which case any namespace is permitted
-				filtered         = false
-				matchingHostname = metav1.ConditionFalse
-				// set to true if ParentRef specifies a Port and a listener matches that Port.
-				portMatched = false
-			)
-			for _, listener := range gateway.Spec.Listeners {
-				// TODO check listenerStatus.SupportedKinds
+		// If the GatewayClass does not match this controller then skip it
+		if gatewayClass.Spec.ControllerName != ControllerName {
+			continue
+		}
 
-				// Check if we already have a matching listener in status.
-				if !existsMatchingReadyListenerInStatus(listener, gateway.Status.Listeners) {
-					continue
-				}
+		// Otherwise we're all set and this controller should reconcile this route.
+
+		var (
+			// Set to true if there exists a listener which wasn't filtered by:
+			// - AlowedRoutes
+			// - listener status checks
+			// - listener and route type checks
+			matched = false
+			// Set to true if ParentRef specified a hostname and it matches route's hostnames.
+			matchingHostname *metav1.ConditionStatus
+			// Set to true if ParentRef specifies a Port and a listener matches that Port.
+			portMatched = false
+
+			allowedByAllowedRoutes  = false
+			allowedBySupportedKinds = false
+		)
+
+		for _, listener := range gateway.Spec.Listeners {
```

Comment:
The logic inside `for` here seems be rather complex. Is it possible to refactor here to extract a function to check whether a listener could match this route?

Reply:
I agree but on the other hand this PR IMO is already an improvement over what has been before. I'd prefer to not go any further in this refactor and treat your suggestion as further improvement that could be done in subsequent PRs. WDYT?

Reply:
I think it is OK to separate the update of logic and refactor into 2 PRs.